lib/rules/prefer-regex-literals.js

```diff
@@ -163,14 +257,83 @@ module.exports = {
                         if (node.arguments.length === 2) {
-                            context.report({ node, messageId: "unexpectedRedundantRegExpWithFlags" });
+                            context.report({
+                                node,
+                                messageId: "unexpectedRedundantRegExpWithFlags",
+                                // eslint-disable-next-line no-loop-func -- scope value won't change
+                                fix(fixer) {
+                                    return fixer.replaceTextRange(node.range, node.arguments[0].raw + getStringValue(node.arguments[1], scope));
+                                }
+                            });
                         } else {
-                            context.report({ node, messageId: "unexpectedRedundantRegExp" });
+                            context.report({
+                                node,
+                                messageId: "unexpectedRedundantRegExp",
+                                fix(fixer) {
+                                    return fixer.replaceTextRange(node.range, node.arguments[0].raw);
+                                }
+                            });
+                        }
+                    } else if (
+                        (getStringValue(node.arguments[0], scope) !== null) &&
+                        (!node.arguments[1] || getStringValue(node.arguments[1], scope) !== null) &&
+                        (node.arguments.length === 1 || node.arguments.length === 2)
+                    ) {
+                        let regexContent = getStringValue(node.arguments[0], scope);
+                        let noFix = false;
+                        let flags;
+
+                        if (node.arguments[1]) {
+                            flags = getStringValue(node.arguments[1], scope);
                         }
-                    } else if (hasOnlyStaticStringArguments(node)) {
-                        context.report({ node, messageId: "unexpectedRegExp" });
+
+                        const RegExpValidatorInstance = new RegExpValidator({ ecmaVersion: getRegexppEcmaVersion(context.parserOptions.ecmaVersion) });
+
+                        try {
+                            RegExpValidatorInstance.validatePattern(regexContent);
+                            if (flags) {
+                                RegExpValidatorInstance.validateFlags(flags);
+                            }
+                        } catch {
+                            noFix = true;
+                        }
+
+                        const tokenBefore = sourceCode.getTokenBefore(node);
+
+                        if (tokenBefore && !validPrecedingTokens.includes(tokenBefore.value)) {
+                            noFix = true;
+                        }
+
+                        if (!/^[-\na-zA-Z\0\r\v\f0-9\\[\](){}\s!@#$%^&*+^_=/~`.><?,'"|:;]*$/u.test(regexContent)) {
```

Comment:
Of course, this is not the best way to solve the [Unicode](https://github.com/eslint/eslint/pull/15077#discussion_r711729114) problem. Is there any better-suggested way?

Reply:
I think this is good enough, but even with this restriction we don't know how these characters appear in the original string literal (unless we want to map each character to its representation in the string literal, but that's overly complex for this feature). For that reason, and other reasons described in https://github.com/eslint/eslint/issues/15029#issuecomment-915961722, I think this feature should be [suggestion](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions), not autofix. It would still have this restriction.

The other question is - do we want to handle all characters from this list? Problematic characters are `/`, `\0` and all whitespace characters except for the space character.